### PR TITLE
Correct the implementation of CRC32

### LINKED ISSUE #363: Possible issue using cx_crc32_hw() when testing on Speculos
Hi,

When testing some CRC32 functionlity using Speculos to emulate a Ledger device the  `cx_crc32_hw()` function seems to be outputting an incorrect CRC.

I have used the following test code to debug the problem and wrote my own `crc32()` function to test and compare against:

````c
uint32_t crc32(const uint8_t *data, size_t len) {
    uint32_t crc = ~0;
    const uint8_t *end = data + len;

    while (data < end) {
        crc ^= *data++;
        for (uint8_t i = 0; i < 8; i++) {
            uint32_t mask = ~((crc & 1) - 1);
            crc = (crc >> 1) ^ (0xEDB88320 & mask);
        }
    }
    return ~crc;
}

// Returns the CRC-32 checksum of the input buffer in network byte order (big endian).
uint32_t crc32_nbo(const uint8_t *bytes, size_t len) {
#if BYTE_ORDER == BIG_ENDIAN
    return crc32(bytes, len);
#elif BYTE_ORDER == LITTLE_ENDIAN
    return os_swap_u32(crc32(bytes, len));
#else
#error "What kind of system is this?"
#endif
}

uint32_t cx_crc32_hw_nbo(const uint8_t *bytes, size_t len) {
#if BYTE_ORDER == BIG_ENDIAN
    return cx_crc32_hw(bytes, len);
#elif BYTE_ORDER == LITTLE_ENDIAN
    return os_swap_u32(cx_crc32_hw(bytes, len));
#else
#error "What kind of system is this?"
#endif
}

void crc_test(void) {
    unsigned char test_data[25] = {0xD9, 0x01, 0x35, 0x55, 0x74, 0x72, 0x00, 0x01, 0x00,
                                   0x41, 0xC0, 0x7B, 0xBC, 0x6A, 0xE0, 0xB7, 0x57, 0xAC,
                                   0x45, 0x93, 0x50, 0xAC, 0x76, 0xE0, 0x51};
    uint32_t crc_1 = 0;
    uint32_t crc_2 = 0;

    crc_1 = cx_crc32_hw_nbo(test_data, 25);
    crc_2 = crc32_nbo(test_data, 25);

    PRINTF("Test data: %.*H\n", 25, test_data);
    PRINTF("CRC from cx_crc32_hw(): %.*H\n", sizeof(crc_1), &crc_1);
    PRINTF("CRC from crc32(): %.*H\n", sizeof(crc_2), &crc_2);
}

````
This outputs the following:
```
10:37:14.417:seproxyhal: printf: Test data: D9013555747200010041C07BBC6AE0B757AC459350AC76E051
10:37:14.419:seproxyhal: printf: CRC from cx_crc32_hw(): E004FC88
10:37:14.421:seproxyhal: printf: CRC from crc32(): 91A6F2D6
```

When checking against several online CRC32 check tools (e.g. https://crccalc.com/) it would seem than my `crc32()`finction outputs the correct CRC and the `cx_crc32_hw()` function is incorrect.

The problem may lie here:

https://github.com/LedgerHQ/speculos/blob/3cf2a3cde7a42605fe1a7b82f6f6c04424259181/src/bolos/cxlib.c#L29-L37

### PR DESCRIPTION
**Pull request by aido**
fixes #363

### CODE REVIEW

File: src/bolos/cx_crc.c [hunk 1]
```diff
@@ -43,7 +43,67 @@ unsigned short sys_cx_crc16_update(unsigned short crc, const void *buf,
...
+  0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
+};
+
+unsigned int sys_cx_crc32_update(unsigned int crc, const void *buf,
+                                   size_t len)
+{
+  const uint8_t *p = buf;
+
+  while (len--)
+    crc = cx_ccitt32[(crc ^ *p++) & 0xff] ^ (crc >> 8u);
```
Comment (xchapron-ledger) [inline src/bolos/cx_crc.c:106]: What do you think of:
```suggestion
  for (i = 0; i < len; i++) {
    crc = cx_ccitt32[(crc ^ *p++) & 0xff] ^ (crc >> 8u);
  }
```
To have explicit `{}` and code equivalent to `sys_cx_crc16_update`

Comment (aido) [reply src/bolos/cx_crc.c]: Yes, I agree. I have changed PR accordingly

**Review by srasoamiaramanana-ledger — APPROVED** [review]

File: src/bolos/cxlib.c [hunk 2]
```diff
@@ -28,10 +28,8 @@ cx_err_t sys_cx_trng_get_random_data(void *buffer, size_t len)
 
 uint32_t sys_cx_crc32_hw(const void *buf, size_t len)
 {
-  // Not a true CRC32, but we don't really need one, here.
   uint32_t crc;
-  crc = sys_cx_crc16_update(0, buf, len);
-  crc |= ((uint32_t)sys_cx_crc16_update((unsigned short)crc, buf, len)) << 16;
+  crc = sys_cx_crc32_update(~0u, buf, len);
```
Comment (xchapron-ledger) [inline src/bolos/cxlib.c:32]:
```suggestion
  crc = sys_cx_crc32_update(0xFFFFFFFF, buf, len);
```
Equivalent but maybe this is easier to understand/see?

**Review by xchapron-ledger — APPROVED** [review]

File: src/bolos/cxlib.c (hunk 2 above)
Comment (aido) [reply src/bolos/cxlib.c:32]: I originally had 0xFFFFFFFF but for some reason changed it to ~0u before committing PR. :-).
 Will change back as you suggest

**Review by xchapron-ledger — DISMISSED** [review]

**Comment by xchapron-ledger** [comment]
@aido Sorry, can you fix the lint: https://github.com/LedgerHQ/speculos/actions/runs/4628356636/jobs/8187724371?pr=364

**Comment by aido** [comment]
> @aido Sorry, can you fix the lint: https://github.com/LedgerHQ/speculos/actions/runs/4628356636/jobs/8187724371?pr=364

No problem. Strange though, lint test fails for the new `sys_cx_crc32_update()` function but not for the existing `sys_cx_crc16_update()` with same layout.

**Comment by xchapron-ledger** [comment]
> > @aido Sorry, can you fix the lint: https://github.com/LedgerHQ/speculos/actions/runs/4628356636/jobs/8187724371?pr=364
> 
> No problem. Strange though, lint test fails for the new `sys_cx_crc32_update()` function but not for the existing `sys_cx_crc16_update()` with same layout.

Same layout, but the return type is a bit shorter, so it ask for a single line... some linter are a bit picky...

File: src/bolos/cx_crc.c [hunk 3]
```diff
@@ -43,7 +43,67 @@ unsigned short sys_cx_crc16_update(unsigned short crc, const void *buf,
...
+unsigned int sys_cx_crc32_update(unsigned int crc, const void *buf, size_t len)
+{
+  const uint8_t *p = buf;
+  size_t i;
+
+  for (i = 0; i < len; i++) {
+    crc = cx_ccitt32[(crc ^ *p++) & 0xff] ^ (crc >> 8u);
+  }
+
+  return crc ^ ~0u;
```
[the remark this answers is not in the document]
Comment (aido) [reply src/bolos/cx_crc.c:108]: Changed. Tested. Working. Done.
Hopefully all is good now for the final push. :-)

**Review by xchapron-ledger — APPROVED** [review]

**Comment by aido** [comment]
A bit off subject but the reason I wanted a working CRC32 in Speculos was to be able to test an application I am writing for Ledger devices.
App is here: https://github.com/aido/app-sskr-check. In your opinion, Is this app a good idea or not, am I wasting my time?

**Comment by xchapron-ledger** [comment]
@aido Thank you for your PR!
Regarding your app, @srasoamiaramanana-ledger probably could answer better than me, but from what she told me, if your app send shares in clear form, the security of the seed is compromised.

**Comment by aido** [comment]
OK, thanks for your comment.

App just displays BIP39 phrases or SSKR phrases on Ledger device display, does not send them anywhere. I may ask @srasoamiaramanana-ledger for their opinion.

**Pull request merged**